module/apmfasthttp/types.go

```diff
@@ -0,0 +1,49 @@
+package apmfasthttp
+
+import (
+	"net/http"
+
+	"github.com/valyala/fasthttp"
+	"go.elastic.co/apm"
+)
+
+type apmHandler struct {
+	requestHandler   fasthttp.RequestHandler
+	tracer           *apm.Tracer
+	requestName      RequestNameFunc
+	requestIgnorer   RequestIgnorerFunc
+	recovery         RecoveryFunc
+	panicPropagation bool
+}
+
+type netHTTPBody struct {
+	b []byte
+}
+
+// Option sets options for tracing server requests.
+type ServerOption func(*apmHandler)
+
+// RequestNameFunc is the type of a function for use in
+// WithServerRequestName.
+type RequestNameFunc func(*fasthttp.RequestCtx) string
+
+// RequestIgnorerFunc is the type of a function for use in
+// WithServerRequestIgnorer.
+type RequestIgnorerFunc func(*fasthttp.RequestCtx) bool
+
+// Transaction wraps the APM transaction.
+// Implements the `io.Closer` interface to end the transaction automatically,
+// due to it will be saved on the RequestCtx.UserValues
+// which will end the transaction automatically when the request finish.
```

Comment:
Also, assuming we don't need the `io.Closer` behaviour - is this `Transaction` wrapper type needed at all?

Reply:
The reason to rely a `io.Closer` is explained at `tx.AutoEndo(...)`:

```go
// AutoEnd defers the tx.End() to call it manually.
//
// So useful, if you need to response in streaming and count too that overtime.
//
// It is enabled by default.
func (tx *Transaction) AutoEnd(v bool)
```

Reply:
This is the main reason to have a transaction wrapper, because when you set a streaming response in `fasthttp`you must set a callback that will be executed after the `RequestHandler` finish, so if you just defer the `tx.End()` as you say, without call `txWrapper.AutoEnd(false)` the transaction will just count the handler spent time but not include the streaming response write.

Reply:
In `net/http`  the `Hijack` and `Streaming` works without return the handler but in `fasthttp` you must set a callback that will executed after all handler/middlewares.

Reply:
Check it out this implementation of `Atreugo` framework: https://github.com/atreugo/apm

Reply:
> This is the main reason to have a transaction wrapper, because when you set a streaming response in fasthttpyou must set a callback that will be executed after the RequestHandler finish, so if you just defer the tx.End() as you say, without call txWrapper.AutoEnd(false) the transaction will just count the handler spent time but not include the streaming response write.

Typically (in other HTTP frameworks) tracing middleware is installed in one location, without needing to modify individual request handlers. If it's necessary to modify streaming response handlers, it will be more difficult to use.

So: is it possible to do this automatically, without forcing the user to call AutoEnd? i.e. detect that a streaming response body has been set, and only end the transaction when it has been fully sent?